File: Android/app/src/main/java/com/social/app/screens/perfil/DataExportManager.kt
package com.social.app.screens.perfil

import com.social.app.backend.SupabaseManager
import io.github.jan.supabase.gotrue.auth
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * "Descargar tus datos" real, comparado con Instagram/Facebook/Twitter-X
 * ("Download Your Information") -- hallazgo real, confirmado con `grep`
 * de "export"/"download_data" sin ningún resultado en todo el repo:
 * `AjustesScreen.kt` ya cubre toggles de privacidad reales (recibos de
 * lectura, última conexión, ubicación...) pero ninguna forma real de
 * exportar los datos propios, pese a que `privacy_policy_es.md` habla de
 * transparencia con los datos del usuario.
 *
 * Alcance deliberado: perfil propio + publicaciones propias + comentarios
 * propios (posts y reels) -- un export representativo, no exhaustivo de
 * cada tabla del esquema (eso sería una ronda propia). Todo filtrado por
 * `auth.uid()` real, reutilizando RLS ya existente (nunca hace falta
 * `service_role`, el propio usuario ya puede leer sus filas).
 */
class DataExportManager {

    @Serializable
    private data class ProfileRow(
        val id: String,
        @SerialName("display_name") val displayName: String,
        val bio: String? = null,
        val username: String? = null,
        @SerialName("created_at") val createdAt: String? = null
    )

    @Serializable
    private data class PostRow(
        val id: String,
        val caption: String? = null,
        @SerialName("media_url") val mediaUrl: String? = null,
        @SerialName("created_at") val createdAt: String = ""
    )

    @Serializable
    private data class CommentRow(
        val id: String,
        val body: String,
        @SerialName("created_at") val createdAt: String = ""
    )

    @Serializable
    private data class ExportPayload(
        val profile: ProfileRow?,
        val posts: List<PostRow>,
        val comments: List<CommentRow>,
        @SerialName("reel_comments") val reelComments: List<CommentRow>,
        @SerialName("exported_at") val exportedAt: String
    )

    /** Devuelve el JSON real ya serializado, o null si no hay sesión. */
    suspend fun buildExportJson(): String? {
        val userId = SupabaseManager.client.auth.currentUserOrNull()?.id ?: return null
        val client = SupabaseManager.client

        val profile = try {
            client.from("profiles")
                .select(columns = Columns.raw("id,display_name,bio,username,created_at")) { filter { eq("id", userId) } }
                .decodeSingleOrNull<ProfileRow>()
        } catch (e: Exception) {
            null
        }
        val posts = try {
            client.from("posts")
                .select(columns = Columns.raw("id,caption,media_url,created_at")) { filter { eq("author_id", userId) } }
                .decodeList<PostRow>()
        } catch (e: Exception) {
            emptyList()
        }
        val comments = try {
            client.from("comments")
                .select(columns = Columns.raw("id,body,created_at")) { filter { eq("author_id", userId) } }
                .decodeList<CommentRow>()
        } catch (e: Exception) {
            emptyList()
        }
        val reelComments = try {
            client.from("reel_comments")
                .select(columns = Columns.raw("id,body,created_at")) { filter { eq("author_id", userId) } }
                .decodeList<CommentRow>()
        } catch (e: Exception) {
            emptyList()
        }

        val payload = ExportPayload(profile, posts, comments, reelComments, java.time.Instant.now().toString())
        return Json { prettyPrint = true }.encodeToString(payload)
    }
}

File: Social/Screens/Perfil/DataExportManager.swift
//
//  DataExportManager.swift
//  Social
//
//  "Descargar tus datos" real, comparado con Instagram/Facebook/
//  Twitter-X ("Download Your Information") -- hallazgo real, confirmado
//  con grep de "export"/"download_data" sin ningún resultado en todo el
//  repo: AjustesView.swift ya cubre toggles de privacidad reales pero
//  ninguna forma real de exportar los datos propios.
//
//  Alcance deliberado: perfil propio + publicaciones propias +
//  comentarios propios (posts y reels) -- un export representativo, no
//  exhaustivo de cada tabla del esquema. Equivalente de
//  DataExportManager.kt.
//

import Foundation

@MainActor
final class DataExportManager: ObservableObject {

    private struct ProfileRow: Encodable, Decodable {
        let id: UUID
        let display_name: String
        let bio: String?
        let username: String?
        let created_at: String?
    }

    private struct PostRow: Encodable, Decodable {
        let id: UUID
        let caption: String?
        let media_url: String?
        let created_at: String
    }

    private struct CommentRow: Encodable, Decodable {
        let id: UUID
        let body: String
        let created_at: String
    }

    private struct ExportPayload: Encodable {
        let profile: ProfileRow?
        let posts: [PostRow]
        let comments: [CommentRow]
        let reel_comments: [CommentRow]
        let exported_at: String
    }

    /// Devuelve una URL real a un fichero temporal ya escrito con el JSON,
    /// o nil si no hay sesión -- listo para `ShareLink(item:)`.
    func buildExportFile() async -> URL? {
        guard let userID = try? await SupabaseManager.shared.client.auth.session.user.id else { return nil }
        let client = SupabaseManager.shared.client

        let profile: ProfileRow? = try? await client
            .from("profiles")
            .select("id,display_name,bio,username,created_at")
            .eq("id", value: userID)
            .single()
            .execute()
            .value

        let posts: [PostRow] = (try? await client
            .from("posts")
            .select("id,caption,media_url,created_at")
            .eq("author_id", value: userID)
            .execute()
            .value) ?? []

        let comments: [CommentRow] = (try? await client
            .from("comments")
            .select("id,body,created_at")
            .eq("author_id", value: userID)
            .execute()
            .value) ?? []

        let reelComments: [CommentRow] = (try? await client
            .from("reel_comments")
            .select("id,body,created_at")
            .eq("author_id", value: userID)
            .execute()
            .value) ?? []

        let payload = ExportPayload(
            profile: profile,
            posts: posts,
            comments: comments,
            reel_comments: reelComments,
            exported_at: ISO8601DateFormatter().string(from: Date())
        )

        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        guard let data = try? encoder.encode(payload) else { return nil }

        let url = FileManager.default.temporaryDirectory.appendingPathComponent("social_datos_\(Int(Date().timeIntervalSince1970)).json")
        do {
            try data.write(to: url)
            return url
        } catch {
            return nil
        }
    }
}
